smart_card_connector_app/src/reader-tracker.js

```diff
@@ -0,0 +1,181 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview This file contains the class used to track readers and their
+ * status based on messages received from the hook setup on PC/SC code inside
+ * NaCl (check readerfactory_nacl.cc for more information).
+ */
+
+goog.provide('GoogleSmartCard.ReaderInfo')
+goog.provide('GoogleSmartCard.ReaderTracker');
+
+goog.require('GoogleSmartCard.DebugDump');
+goog.require('GoogleSmartCard.Logging');
+goog.require('GoogleSmartCard.TypedMessage');
+goog.require('goog.asserts');
+goog.require('goog.json.NativeJsonProcessor');
+goog.require('goog.log.Logger');
+goog.require('goog.messaging.AbstractChannel');
+goog.require('goog.object');
+goog.require('goog.structs.Map');
+
+goog.scope(function() {
+
+/** @const */
+var GSC = GoogleSmartCard;
+
+/**
+ * Structure used to store information about the reader.
+ * @param {string} name
+ * @param {string=} status
+ * @param {string=} error
+ * @constructor
+ * @struct
+ */
+GSC.ReaderInfo = function(name, status, error) {
+  this.name = name;
+  this.status = status;
+  this.error = error;
+};
+
+/** @const */
+var ReaderInfo = GSC.ReaderInfo;
+
+/**
+ * This class tracks readers, provides methods to retrieve the list of readers
+ * and get updates on change, and listens to messages received from the hook
+ * living in the NaCl space.
+ * @param {!goog.messaging.AbstractChannel} messageChannel
+ * @param {!goog.log.Logger} parentLogger
+ * @constructor
+ */
+GSC.ReaderTracker = function(messageChannel, parentLogger) {
+  /** @private */
+  this.logger_ = parentLogger;
+
+  messageChannel.registerService(
+      'reader_init_add', this.readerInitAddListener_.bind(this));
+  messageChannel.registerService(
+      'reader_finish_add', this.readerFinishAddListener_.bind(this));
+  messageChannel.registerService(
+      'reader_remove', this.readerRemoveListener_.bind(this));
+
+  /**
+   * @type {!goog.structs.Map.<string, !GSC.ReaderInfo>}
+   * @private
+   */
+  this.readers_ = new goog.structs.Map;
+
+  /**
+   * @type {!Array.<!function()>}
+   * @private
+   */
+  this.onUpdateListeners_ = [];
+
+  this.logger_.fine('Initialized');
+};
+
+/** @const */
+var ReaderTracker = GSC.ReaderTracker;
+
+/**
+ * TODO: I need to figure out the param because the closure compiler complains
+ *       about mismatched types, will do it at a later time.
+ * param {Object|string} message
+ * @private
+ */
+ReaderTracker.prototype.readerInitAddListener_ = function(message) {
+  this.logger_.warning('readerInitAddListener_ ' + message);
```

Comment:
When working on the final implementation, please make sure to produce meaningful logs with the "info" level. Because that's very useful information for investigating the issues post-factum.